### PR TITLE
Fix enabled state for edit line box

### PR DESCRIPTION
**Pull request by PragTob**
This includes 4 separate changes related to the state of elements, for the changes themselves I'll refer you to the commit messages :)

Fixes (most of) #1183

### CODE REVIEW

File: shoes-core/lib/shoes/common/state.rb [hunk 1]
```diff
@@ -1,16 +1,30 @@
 class Shoes
   module Common
     module State
```
Comment (PragTob) [inline shoes-core/lib/shoes/common/state.rb]: curiously we had this, but it wasn't included anywhere o_O might have gotten lost in the big style PR or somewhere - another reason against big PRs :D

File: shoes-swt/lib/shoes/swt/input_box.rb [hunk 2]
```diff
@@ -32,7 +32,7 @@ def text=(value)
       end
 
       def enabled(value)
-        @real.enable_widget value
+        @real.enabled = value
```
Comment (PragTob) [inline shoes-swt/lib/shoes/swt/input_box.rb]: this change is a bit unrelated, I couldn't find any reference to `enableWidget` in the SWT docs so I changed it to the documented enabled

File: shoes-core/spec/shoes/shared_examples/state.rb [hunk 3]
```diff
@@ -1,7 +1,8 @@
 shared_examples_for "object with state" do
   let(:input_opts) { {state: "disabled"} }
 
-  it "should initialize" do
+  it "should initialize in the right state" do
+    expect_any_instance_of(Shoes.configuration.backend_class(described_class)).to receive(:enabled).with(false)
```
Comment (jasonrclark) [inline shoes-core/spec/shoes/shared_examples/state.rb]: Was this added just to test that the initialization calls through to the backend?

**Comment by jasonrclark** [comment]
Had one question just to clarify about specs, but otherwise this looks :sparkling_heart: Thanks for getting to this @PragTob! I'll get the CHANGELOG bumped once we merge this.

File: shoes-core/spec/shoes/shared_examples/state.rb (hunk 3 above)
Comment (PragTob) [inline shoes-core/spec/shoes/shared_examples/state.rb]: Yep, indeed it was. I.e. testing the `after_initialize` bit. Proved its worth, i.e. The one in list_box didn't call super, hence did not call the `State` module method and hence it failed :)

Comment (jasonrclark) [inline shoes-core/spec/shoes/shared_examples/state.rb]: :+1:

Comment (PragTob) [inline shoes-core/spec/shoes/shared_examples/state.rb]: I'd have loved for this to be tested on the GUI instance and not as `Shotgun Surgery` on any class, but I couldn't find a suitable way since the gui is constructed in initialize and then the method is straight up called.

There might be a way, now that I think about it again - stubbing the method that creates the gui the return a double. But it should be fine as is imo :)

**Comment by PragTob** [comment]
:) felt great to do some coding again, also this was an easy one. Super happy with how little code I needed to change for these changes and how easy it was.

**Comment by jasonrclark** [comment]
Huzzah! :tada:

File: shoes-core/lib/shoes/common/state.rb (hunk 1 above)
Comment (KCErb) [inline shoes-core/lib/shoes/common/state.rb]: I found where I did this:

https://github.com/shoes/shoes4/pull/803/files#r15267351

 Not sure I understand my reasoning then . . . what we have here seems good :+1:

Comment (PragTob) [inline shoes-core/lib/shoes/common/state.rb]: Cool that you dug that up!

I guess the reasoning was most methods are auto generated by style, which is true. But now we needed more stuff to reinstate it - software always changes and it is not uncommon to go back and forth on a decision (at least in my experience) :)
